packages/synapse-react-client/src/utils/hooks/usePreFetchResource.ts

```diff
@@ -18,35 +19,36 @@ export function releaseResourceUrl(resourceUrl: string) {
 export default function usePreFetchResource(
   preSignedURL?: string,
 ): string | undefined {
-  const [blob, setBlob] = useState<Blob | undefined>(undefined)
-  const url = useCreateUrlForData(blob)
+  const { data: blob, error } = useQuery(
+    ['usePreFetchResource', preSignedURL],
+    () => fetchBlob(preSignedURL!),
+    {
+      enabled: !!preSignedURL,
+      // The URL may expire, so the fetched item should never be marked as 'stale'; a refetch may not work.
+      staleTime: Infinity,
+    },
+  )
 
   useEffect(() => {
-    let isMounted = true
-    const getData = async (url: string) => {
-      try {
-        const blob = await fetchBlob(url)
-        if (isMounted) {
-          setBlob(blob)
-        }
-      } catch (e) {
-        console.error(
-          `Failed to fetch object with presigned URL ${url}. See network log for details`,
-        )
-      }
-    }
-    if (preSignedURL) {
-      getData(preSignedURL)
+    if (error) {
+      console.error(
+        `Failed to fetch object with presigned URL ${preSignedURL}. See network log for details`,
+      )
     }
+  }, [error, preSignedURL])
 
-    return () => {
-      isMounted = false
-    }
-  }, [preSignedURL])
-
-  return url
+  // While the retrieved data is cached by react-query, the object created by URL.createObjectURL is not. More specifically,
+  // the stored object is different for each instance of this hook, so when one instance tears down and deletes its object,
+  // other instances of the hook are not affected.
+  // This uses more memory than necessary in cases where the same object is fetched multiple times, but it's a tradeoff
+  // for simple cache and resource management.
+  return useCreateUrlForData(blob)
```

Comment:
The old behavior was that with _n_ instances of this hook with the same URL, you would send _n_ HTTP requests and have _n_ objects stored in memory.

With react-query we now have 1 HTTP request, but still storing _n_ things in memory. I don't think the storage issue is causing any problems today. I also think that solution would add complexity in a way that's hard to get right, so I'm intentionally keeping unoptimal behavior to avoid introducing a new error.

We'd probably need to revisit this if we start downloading files larger than profile pictures with this hook.